cms/lib/xblock/upstream_sync.py

```diff
@@ -138,201 +144,81 @@ def try_get_for_block(cls, downstream: XBlock) -> t.Self:
         """
+        # We import this here b/c UpstreamSyncMixin is used by cms/envs, which loads before the djangoapps are ready.
+        from openedx.core.djangoapps.content_libraries import api as lib_api
+
         if not downstream.upstream:
             raise NoUpstream()
         if not isinstance(downstream.usage_key.context_key, CourseKey):
             raise BadDownstream(_("Cannot update content because it does not belong to a course."))
-        if downstream.has_children:
-            raise BadDownstream(_("Updating content with children is not yet supported."))
+        downstream_type = downstream.usage_key.block_type
+
+        upstream_key: LibraryUsageLocatorV2 | LibraryContainerLocator
         try:
             upstream_key = LibraryUsageLocatorV2.from_string(downstream.upstream)
-        except InvalidKeyError as exc:
-            raise BadUpstream(_("Reference to linked library item is malformed")) from exc
-        downstream_type = downstream.usage_key.block_type
-        if upstream_key.block_type != downstream_type:
-            # Note: Currently, we strictly enforce that the downstream and upstream block_types must exactly match.
-            #       It could be reasonable to relax this requirement in the future if there's product need for it.
-            #       For example, there's no reason that a StaticTabBlock couldn't take updates from an HtmlBlock.
+        except InvalidKeyError:
+            try:
+                upstream_key = LibraryContainerLocator.from_string(downstream.upstream)
+            except InvalidKeyError as exc:
+                raise BadUpstream(_("Reference to linked library item is malformed")) from exc
+
+        if isinstance(upstream_key, LibraryUsageLocatorV2):
+            # The upstream is an XBlock
+            if downstream.has_children:
+                raise BadDownstream(
+                    _("Updating content with children is not yet supported unless the upstream is a container."),
+                )
+            expected_downstream_block_type = upstream_key.block_type
+            try:
+                block_meta = lib_api.get_library_block(upstream_key)
+            except XBlockNotFoundError as exc:
+                raise BadUpstream(_("Linked upstream library block was not found in the system")) from exc
+            version_available = block_meta.published_version_num
+        else:
+            # The upstream is a Container:
+            assert isinstance(upstream_key, LibraryContainerLocator)
+            try:
+                container_meta = lib_api.get_container(upstream_key)
+            except lib_api.ContentLibraryContainerNotFound as exc:
+                raise BadUpstream(_("Linked upstream library container was not found in the system")) from exc
+            expected_downstream_block_type = container_meta.container_type.olx_tag
+            version_available = container_meta.published_version_num
+
+        if downstream_type != expected_downstream_block_type:
```

Comment:
I'm getting an error here when I try to add a library Unit to a course, because:

```
downstream_type == 'unit'
expected_downstream_block_type == 'vertical'
```

Is the downstream block being created with the wrong block type? Or should this be this instead (so that `expected_downstream_block_type  == 'unit'` too)?

```suggestion
            expected_downstream_block_type = container_meta.container_type.value
            version_available = container_meta.published_version_num

        if downstream_type != expected_downstream_block_type:
```

Reply:
@pomegranited You need to specify `category: "vertical"` in the API call. See Braden's comment above:

> Note that I think the "Manual Test instructions" are out of date; I have mostly been working on the integration tests in test_downstream_sync_integration.py. But specifically, the API request needs to have category: vertical even though it's a unit, and I don't think it needs "type":"library_v2". See comments in test_downstream_sync_integration.py for details.

Reply:
Yep - I mentioned this in the test case as well:

```
# The API consumer needs to specify "vertical" here, even though upstream is "unit".
# In the future we could create a nicer REST API endpoint for this that's not part of
# the messy '/xblock/' API and which auto-detects the types based on the upstream_key.
```

@navinkarkera had some code to modify the `/xblock/` endpoint to override this automatically, but I'm a little cautious about changing that existing API endpoint too much (since `unit` _is_ also a valid XBlock some people might want to create - we used it for LabXchange). So I'd much rather we make a new API in the future for this thing, like `/api/contentstore/v2/downstreams/clone/` or something like that, and the new endpoint can auto-set `vertical` for you based on the source/upstream container type, using the nice new code in https://github.com/openedx/edx-platform/pull/36580